clang/test/Sema/code_align.c

```diff
@@ -0,0 +1,126 @@
+  [[clang::code_align(4)]]
+  i = 7;
+  for (i = 0; i < 10; ++i) {
+    a[i] = b[i] = 0;
+  }
+
+  // expected-error@+1{{'code_align' attribute cannot be applied to a declaration}}
+  [[clang::code_align(12)]] int n[10];
+}
+
+void bar(int);
+// cpp-local-note@+1{{declared here}}
+void foo1(int A)
+{
+  // expected-error@+1{{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was 0}}
+  [[clang::code_align(0)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // expected-error@+1{{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was -4}}
+  [[clang::code_align(-4)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+    // cpp-local-error@+2{{integral constant expression must have integral or unscoped enumeration type, not 'double'}}
+    // c-local-error@+1{{integer constant expression must have integer type, not 'double'}}
+  [[clang::code_align(64.0)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // expected-error@+1{{'code_align' attribute takes one argument}}
+  [[clang::code_align()]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // expected-error@+1{{'code_align' attribute takes one argument}}
+  [[clang::code_align(4,8)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // no diagnostic is expected
+  [[clang::code_align(32)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // cpp-local-error@+2{{integral constant expression must have integral or unscoped enumeration type, not 'const char[4]'}}
+  // c-local-error@+1{{integer constant expression must have integer type, not 'char[4]'}}
+  [[clang::code_align("abc")]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  [[clang::code_align(64)]] // expected-note{{previous attribute is here}}
+  [[clang::code_align(64)]] // expected-error{{duplicate loop attribute 'code_align'}}
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // expected-error@+1{{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was 7}}
+  [[clang::code_align(7)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // expected-error@+1{{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was 5000}}
+  [[clang::code_align(5000)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // expected-warning@+2 {{integer literal is too large to be represented in a signed integer type, interpreting as unsigned}}
+  // expected-error@+1 {{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was 9223372036854775808}}
+  [[clang::code_align(9223372036854775808)]]
+  for(int I=0; I<256; ++I) { bar(I); }
+
+  // expected-error@+1 {{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was 0}}
```

Comment:
This looks like a bug to me.  Why does it say 0?

Reply:
@erichkeane ,This is causing assertion instead of diagnostic
Assertion failed: getActiveBits() <= 64 && "Too many bits for uint64_t", file C:\ws\src\llvm\include\llvm/ADT/APInt.h, line 1488
https://buildkite.com/llvm-project/clang-ci/builds/6708#018bcc89-82bd-49de-8b33-1e210d1a48ad

Reply:
Right, that is more or less what I was saying before, if you're converting an `APInt` that is too large for conversion operators, you'll end up getting the wrong answer or asserting, etc.  So you have to manage how that comes out better.

Reply:
@erichkeane, i have fixed the issue with commit https://github.com/llvm/llvm-project/pull/70762/commits/8bec3c455b5f55b40a8a3d8c9180c0e3d52765fa

Now This is returning a better value:
'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was 2147483647
This represents the value of the provided argument truncated to an int (instead of a 128bit float).